### PR TITLE
pythonPackages.glom: init at 18.3.1

### PR DESCRIPTION
**Pull request by Twey**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ma27** [comment]
as you've added quite a bunch of new python packages: do you intend to maintain these packages in the future (if so, please add the `meta.maintainers` field as well) :)

**Comment by Twey** [comment]
@Ma27 Ah, thanks for reminding me >.<

Requested changes made. :)

**Review by costrouc — APPROVED** [review]

**Comment by costrouc** [comment]
@GrahamcOfBorg build python27Packages.boltons  python27Packages.face  python27Packages.glom  python37Packages.boltons  python37Packages.face  python37Packages.glom

File: pkgs/development/python-modules/face/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0zdp5qlrhxf4dypvvd0zr7zxj2svkz9wblp37vgw01wvcy9b1ds7";
+  };
+
+  propagatedBuildInputs = [ boltons ];
+
+  checkInputs = [ pytest ];
+  checkPhase = "py.test face/test";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/face/default.nix]: Why not use `pytest`?

File: pkgs/development/python-modules/glom/default.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "10rabxqg6v03mcsavdpd5j3slvxmmxj64fn4409zmn12sb0l4a6v";
+  };
+
+  propagatedBuildInputs = [ boltons attrs face ];
+
+  checkInputs = [ pytest ];
+  checkPhase = "py.test glom/test";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/glom/default.nix]: Why not use `pytest`?

**Comment by worldofpeace** [comment]
`glom` is also a cli tool. Is that meant to be used exclusively for certain purposes?

If so, we need to add an attribute to `all-packages.nix` using `toPythonApplication`.
A better explanation can be seen in the python section of the nixpkgs manual.

**Comment by worldofpeace** [comment]
Additionally I'm seeing `Ran 0 tests in 0.000s` in the output when building `boltons`.
So I don't think tests are being ran.

File: pkgs/development/python-modules/boltons/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  version = "18.0.1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1l845lh6gr5mpwcif8qcxksvira7ksr9cipcz5wpk7ll5s4x6z8v";
+  };
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/mahmoud/boltons;
+    description = "Functionality that should be in the standard library";
```
[the remark this answers is not in the document]
Comment (Twey) [reply pkgs/development/python-modules/boltons/default.nix]: Where's that coming from? I usually try to use (some subset of) the description from PyPI.

[the remark this answers is not in the document]
Comment (worldofpeace) [reply pkgs/development/python-modules/boltons/default.nix]: The title on github

**Comment by Twey** [comment]
> Additionally I'm seeing `Ran 0 tests in 0.000s` in the output when building `boltons`.
> So I don't think tests are being ran.

The version of `boltons` on PyPI doesn't include any tests (or the requirements file required to even run the test command).

**Comment by Twey** [comment]
Is that wise? Packages that rely on the PyPI version won't be able to use it as a dependency…

**Comment by Ma27** [comment]
> Is that wise? Packages that rely on the PyPI version won't be able to use it as a dependency…

This happens for relatively many packages. During the `setup.py install` phase the `version` field in `setup.py` of all packages in `$PYTHONPATH` will be parsed and checked if they're compatible.

Or am I misunderstanding your concern?

**Comment by Twey** [comment]
All requested changes made!  How's it looking?

**Comment by Twey** [comment]
Build fails because 2b9283afc5bccdd06f60720064e492f076ee9a37 introduced a different version of glom in the meantime (as an application, but without the Python library).  Please advise.

**Comment by Twey** [comment]
Scratch that, it's actually a different application, we just have a name conflict.

**Comment by Twey** [comment]
Have renamed `glom` to `glom-db` pending input from @jtojnar.

**Comment by jtojnar** [comment]
Looking at [repology](https://repology.org/metapackages/?search=glom) `glom` package name seems to be used for the DB tool. I would prefer to remain consistent with other distros.

**Comment by Twey** [comment]
I'd prefer it too, but we have strict rules on naming Python packages and less strict guidelines on other applications. I'm open to alternative suggestions though.

**Comment by worldofpeace** [comment]
> I'm also okay with not packaging the application (at least until this is resolved), but perhaps @worldofpeace would be unhappy with that solution?

This is glom's docs on their cli tool https://glom.readthedocs.io/en/latest/cli.html
> glom’s CLI is still under construction. Definitely usable and useful, but glom is a library first, and if you’re reading this, the CLI should not be considered stable.

I honestly don't think it's that important at this point to have it as an application in `all-packages.nix`.
@Ma27 Any thoughts?

File: pkgs/development/python-modules/boltons/default.nix [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{ stdenv, buildPythonPackage, fetchFromGitHub, pytest }:
```
Comment (worldofpeace) [inline pkgs/development/python-modules/boltons/default.nix]:
```suggestion
{ stdenv, buildPythonPackage, fetchFromGitHub, pytest }:

```

File: pkgs/development/python-modules/boltons/default.nix [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+buildPythonPackage rec {
+  pname = "boltons";
+  version = "2018-12-31";
```
Comment (worldofpeace) [inline pkgs/development/python-modules/boltons/default.nix]: I think the date of this commit is different https://github.com/mahmoud/boltons/commit/62fa806d9c86daf3f8c23e807cd853db2567b179

Also why not use the a release https://github.com/mahmoud/boltons/releases/tag/18.0.1?

**Comment by Ma27** [comment]
> I honestly don't think it's that important at this point to have it as an application in all-packages.nix.
@Ma27 Any thoughts?

As the package states to be a "library first" I personally think that no top-level alias is needed, especially if it's still "under construction" (an "unstable" CLI is acceptable IMHO, but "under construction" is even worse).

IIRC the top-level alias makes sense if you want to package an application which happens to be python based and is not intended to be a library, but correct me if I'm wrong :D

File: pkgs/development/python-modules/boltons/default.nix (hunk 5 above)
Comment (Twey) [reply pkgs/development/python-modules/boltons/default.nix]: Release doesn't test :(  Date updated, and bumped the commit whilst I was at it.

**Comment by Twey** [comment]
Okay, removed the application.

**Comment by worldofpeace** [comment]
@GrahamcOfBorg build pythonPackages.boltons python3Packages.boltons pythonPackages.face python3Packages.face pythonPackages.glom python3Packages.glom

**Comment by worldofpeace** [comment]
I think this is good, I'm going for it.

**Pull request merged**